src/main/drivers/pwm_output.c

```diff
@@ -231,27 +192,87 @@ void pwmFastPwmMotorConfig(const timerHardware_t *timerHardware, uint8_t motorIn
         timerMhzCounter = MULTISHOT_TIMER_MHZ;
         pwmWritePtr = pwmWriteMultiShot;
         break;
+    case (PWM_TYPE_BRUSHED):
+        timerMhzCounter = PWM_BRUSHED_TIMER_MHZ;
+        pwmWritePtr = pwmWriteBrushed;
+        motorConfig->use_unsyncedPwm = true;
+        idlePulse = 0;
+        break;
+    case (PWM_TYPE_STANDARD):
+        timerMhzCounter = PWM_TIMER_MHZ;
+        pwmWritePtr = pwmWriteStandard;
+        motorConfig->use_unsyncedPwm = true;
+        idlePulse = 0;
+        break;
     }
+    
+    for (int motorIndex = 0; motorIndex < MAX_SUPPORTED_MOTORS && motorIndex < motorCount; motorIndex++) {
+        ioTag_t tag = motorConfig->motorTags[motorIndex];
+        
+        if (DEFIO_TAG_ISEMPTY(tag)) {
+            break;
+        }
 
-    if (motorPwmRate > 0) {
-        const uint32_t hz = timerMhzCounter * 1000000;
-        motors[motorIndex] = pwmOutConfig(timerHardware, timerMhzCounter, hz / motorPwmRate, idlePulse);
-    } else {
-        motors[motorIndex] = pwmOutConfig(timerHardware, timerMhzCounter, 0xFFFF, 0);
+        motors[motorIndex].io = IOGetByTag(tag);
+        
+        IOInit(motors[motorIndex].io, OWNER_MOTOR, RESOURCE_OUTPUT, RESOURCE_INDEX(motorIndex));
+        IOConfigGPIO(motors[motorIndex].io, IOCFG_AF_PP);
+        
+        const timerHardware_t *timer = timerGetByTag(tag, TIMER_OUTPUT_ENABLED);
+        
+        if (timer == NULL) {
+            /* TODO: flag failure and disable ability to arm */
+            break;
+        }
+        
+        motors[motorIndex].pwmWritePtr = pwmWritePtr;
+        if (motorConfig->use_unsyncedPwm) {
+            const uint32_t hz = timerMhzCounter * 1000000;
+            pwmOutConfig(&motors[motorIndex], timer, timerMhzCounter, hz / motorConfig->motor_pwm_rate, idlePulse);
+        } else {
+            pwmOutConfig(&motors[motorIndex], timer, timerMhzCounter, 0xFFFF, 0);
+        }
+        motors[motorIndex].enabled = true;
     }
-    motors[motorIndex]->pwmWritePtr = pwmWritePtr;
 }
 
-#ifdef USE_SERVOS
-void pwmServoConfig(const timerHardware_t *timerHardware, uint8_t servoIndex, uint16_t servoPwmRate, uint16_t servoCenterPulse)
+pwmOutputPort_t *pwmGetMotors()
 {
-    servos[servoIndex] = pwmOutConfig(timerHardware, PWM_TIMER_MHZ, 1000000 / servoPwmRate, servoCenterPulse);
+    return motors;
 }
 
+#ifdef USE_SERVOS
 void pwmWriteServo(uint8_t index, uint16_t value)
 {
-    if (servos[index] && index < MAX_SERVOS) {
-        *servos[index]->ccr = value;
+    if (index < MAX_SUPPORTED_SERVOS && servos[index].ccr) {
+        *servos[index].ccr = value;
     }
 }
+
+void servoInit(escAndServoConfig_t *servoConfig)
+{
+    for (uint8_t servoIndex = 0; servoIndex < MAX_SUPPORTED_SERVOS; servoIndex++) {
+        ioTag_t tag = servoConfig->servoTags[servoIndex];
+        
+        if (DEFIO_TAG_ISEMPTY(tag)) {
```

Comment:
(not a review, just browsing the changes)

IO code was designed to guarantee that `if(ioTag) { /* IO tag is not empty */}` and `if(io) { /* io refers to pin */ }` works as expected.
`DEFIO` macros are semi-internal - most code should be fine without them. `IO_TAG` is preferred interface to convert  IO name to tag. `DEFIO_REC` was created when I implemented IO code to get record directly (can be used to initialize global variable etc), but currently I prefer to use `IOGetByTag` when possible